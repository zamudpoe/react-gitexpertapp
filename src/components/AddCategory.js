import React, { useState } from 'react'
import PropTypes from "prop-types"

export const AddCategory = ( { setCategories } ) => {
  const [ inputValue, setInputValue ] = useState("")

  const handleInputChange             = ( e ) => { 
    setInputValue(e.target.value) 

    console.log("HandleInputChange llamado");

  }

  const handleSubmit                  = ( e ) => {  
    e.preventDefault() 
    console.log("%chandleSubmit de <AddCategory /> %s","color: coral; font-size: 1.2rem;", inputValue)

    if ( inputValue.trim().length > 2 ) {  
      setCategories( categories => [ inputValue, ...categories ] )
      /* setCategories( 12345 ) */ // forzamos que falle la prueba con esta linea y comentando la de arriba
      setInputValue("")
    } else {  
      console.log("%cPor favor dame mas de dos letras!","color: tomato;")  
    }  
  }

  return (
    <form onSubmit= { handleSubmit }>
      <h1>{ inputValue }</h1>
      <input 
        type        = "text" 
        value       = { inputValue }
        onChange    = { handleInputChange }        
        onSubmit    = { handleSubmit }
        placeholder = "Introduce un valor"
      />
    </form>
  ) 
}


AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}

