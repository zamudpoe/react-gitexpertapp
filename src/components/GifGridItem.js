import React from 'react'
import PropTypes from 'prop-types'

export const GifGridItem = ( { id, title, url } ) => {

  /* console.log( { id, title, img } ) */
  /* https://animate.style/ */
  return (
    <div className="card animate__animated animate__zoomInDown" key={ id }>
      <img src={ url } alt={ title } />
      <p> { title } </p>
    </div>
  )

}


/* 
  1. Enzyme
    npm i --save-dev enzyme enzyme-adapter-react-16
  2. Enzyme to JSON
    
  3. PRUEBAS
    *.- Debe moistrar el componente correctamente!.
      *.- Shallow 
      *.- Wrapper
      *- wrapper .toMatchSnapshot()

*/

GifGridItem.propTypes = {
  title : PropTypes.string.isRequired, 
  url   : PropTypes.string.isRequired 
}

