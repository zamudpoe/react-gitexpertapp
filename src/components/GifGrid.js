/* import React, { useState, useEffect, Fragment } from 'react' */
import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem'

export const GifGrid = ( { category } ) => {  
  const { data:images, loading } = useFetchGifs(category)
  /* console.log(data, loading) */

  return ( 

    <>
      <h3>{ category }</h3> <hr />      

      { loading && <p>Loading...</p> }

      {/* 04 destructuramos el estado y lo enviamos al nuevo componente GifGridItem */}
      <div className="card-grid"> 

        {
          images.map( img => ( 
            <GifGridItem 
              key   = { img.id }  
              { ...img }
            />
          ))
        }

      </div>  
    </>

  ) 
}  


