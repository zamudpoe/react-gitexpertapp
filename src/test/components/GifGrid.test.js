import React from "react";
import '@testing-library/jest-dom';

import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";

jest.mock( '../../hooks/useFetchGifs' )  // nos ayuda a fingir datos 

describe('Pruebas en el <GifGrid />', () => {
  const category = "Hulk"
  
  // Prueba 01
  test('Debe de mostrar <GifGrid /> correctamente!', ( ) => {

    /* FALSEAMOS la data con mockReturnValue */
    useFetchGifs.mockReturnValue({
      data    : [], 
      loading : true 
    })
    
    const wrapper  = shallow( <GifGrid category = { category } /> )
    console.log("Loading: %s", wrapper.html().includes("Loading...") )
    expect( wrapper ).toMatchSnapshot()
  })

  // Prueba 02
  test('Debe de mostrar items cuando se cargan imagenes useFetchGifs', () => {

    const gifs = [{
      id       : "100",
      url      : "https://localhost:3000",
      title    : "Titulo XXX"
    }, {
      id       : "200",
      url      : "https://localhost:5500",
      title    : "Mama esta presa"
    }]

    /* FALSEAMOS la data con mockReturnValue */
    useFetchGifs.mockReturnValue({
      data    : gifs, 
      loading : false 
    })

    const wrapper  = shallow( <GifGrid category = { category } /> )

    expect( wrapper ).toMatchSnapshot()
    expect( wrapper.find('p').exists() ).toBe( false ) 
    expect( wrapper.find("GifGridItem").length ).toBe( gifs.length )

  });

});


