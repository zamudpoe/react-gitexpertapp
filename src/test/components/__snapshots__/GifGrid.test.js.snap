// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <GifGrid /> Debe de mostrar <GifGrid /> correctamente! 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Hulk
  </h3>
   
  <hr />
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Pruebas en el <GifGrid /> Debe de mostrar items cuando se cargan imagenes useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Hulk
  </h3>
   
  <hr />
  <div
    className="card-grid"
  >
    <GifGridItem
      id="100"
      key="100"
      title="Titulo XXX"
      url="https://localhost:3000"
    />
    <GifGridItem
      id="200"
      key="200"
      title="Mama esta presa"
      url="https://localhost:5500"
    />
  </div>
</Fragment>
`;
